[ews-build.webkit.org] Specify branches to trigger queues on

JonWBedard opened the pull request:
#### ae57f1da7af0f7c3b2aab1ae26084550b5d52879
<pre>
[ews-build.webkit.org] Specify branches to trigger queues on
<a href="https://bugs.webkit.org/show_bug.cgi?id=255562">https://bugs.webkit.org/show_bug.cgi?id=255562</a>
rdar://108168587

Reviewed by Aakash Jain.

Some queues should only be triggered by certain branches. Allow factories
to define this.

* Tools/CISupport/ews-build/factories.py:
(Factory): Add &apos;branches&apos; class variable.
(Factory.__init__): Pass &apos;branches&apos; to ValidateChange.
(WinCairoFactory): Only run WinCairo tests on &apos;main&apos;.
(GTKBuildFactory): Only run tests on &apos;main&apos; and &apos;webkit&apos; branches.
(WPEBuildFactory): Ditto.
* Tools/CISupport/ews-build/steps.py:
(ValidateChange.__init__): Allow caller to limit queue to specific branches.
(ValidateChange.run): Skip build if target branch is not supported by this queue.
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/267047@main">https://commits.webkit.org/267047@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7129174bb3a1c2fc6357d2618a47b11a2b2e637e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/15488 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/15793 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/16156 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/17243 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/14523 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/15660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/18310 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/15888 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/17243 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/15673 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/18310 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/16156 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/17986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/23/builds/18310 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/16156 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/17986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/18310 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/16156 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/17986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/14719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/15888 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| [  ~~🧪 services~~](https://ews-build.webkit.org/#/builders/28/builds/15387 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/13976 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/16156 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/18337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1886 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/14539 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aj062 reviewed (changes requested):
As discussed on Slack earlier, GitHub Status-bubbles will always keep showing waiting for those queues in this implementation.
Maybe for that reason, it's better to still trigger every queue and skip builds explicitly (using step like validate-change or something).

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1694,6 +1699,11 @@ def fail_build(self, reason):
         pr_number = self.getProperty('github.number', '')
+        branch = self.getProperty('github.base.ref', DEFAULT_BRANCH)
+
+        if not any(candidate.match(branch) for candidate in self.branches):
```
Please ensure that this doesn't break patch workflow.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1694,6 +1699,11 @@ def fail_build(self, reason):
+        branch = self.getProperty('github.base.ref', DEFAULT_BRANCH)
+
+        if not any(candidate.match(branch) for candidate in self.branches):
+            rc = yield self.skip_build(f"Changes to '{branch}' are not tested")
```
Nit: "are not tested" => "are not tested by EWS"

aj062 reviewed (approved):
r+ assuming it is tested on uat instance or locally.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1664,6 +1665,10 @@ def __init__(
         self.enableSkipEWSLabel = enableSkipEWSLabel
         self.addURLs = addURLs
+
+        branches = branches or [r'.+']
```
Maybe .* instead of .+ to handle accidental case when branch is "".

JonWBedard committed:
[ews-build.webkit.org] Specify branches to trigger queues on
https://bugs.webkit.org/show_bug.cgi?id=255562
rdar://108168587

Reviewed by Aakash Jain.

Some queues should only be triggered by certain branches. Allow factories
to define this.

* Tools/CISupport/ews-build/factories.py:
(Factory): Add 'branches' class variable.
(Factory.__init__): Pass 'branches' to ValidateChange.
(WinCairoFactory): Only run WinCairo tests on 'main'.
(GTKBuildFactory): Only run tests on 'main' and 'webkit' branches.
(WPEBuildFactory): Ditto.
* Tools/CISupport/ews-build/steps.py:
(ValidateChange.__init__): Allow caller to limit queue to specific branches.
(ValidateChange.run): Skip build if target branch is not supported by this queue.
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: https://commits.webkit.org/267047@main
